ddtrace/contrib/starlette/patch.py

```diff
@@ -65,3 +53,49 @@ def unpatch():
     setattr(starlette, "_datadog_patch", False)
 
     _u(starlette.applications.Starlette, "__init__")
+
+    # We need to check that Fastapi instrumentation hasn't already unpatched these
+    if isinstance(starlette.routing.Route.handle, ObjectProxy):
+        _u(starlette.routing.Route, "handle")
+
+    if isinstance(starlette.routing.Mount.handle, ObjectProxy):
+        _u(starlette.routing.Mount, "handle")
+
+
+def traced_handler(wrapped, instance, args, kwargs):
+    # Since handle can be called multiple times for one request, we take the path of each instance
+    # Then combine them at the end to get the correct resource name
+    scope = get_argument_value(args, kwargs, 0, "scope")
+
+    if "__dd_paths__" in scope:
+        scope["__dd_paths__"].append(instance.path)
+
+    else:
+        scope["__dd_paths__"] = [instance.path]
+
+    if "datadog" not in scope:
+        return wrapped(*args, **kwargs)
+
+    # The scope["datadog"] object gets replaced with each new call to the handler
```

Comment:
@Kyle-Verhoog  what do you think about this implementation. Part of the logic revolves around the fact that the `scope["datadog"]` object gets replaced with each new call to the handler. Therefore we can't just pull from that once and move on. Instead we need to take the first span and put it somewhere else in the `scope` so it doesn't get replaced once a new handler is hit.

Reply:
hmm is this because fastapi/starlette call back into the app for the subapps?

What we could do is only set the span in the ASGI scope once (if there isn't a span in the scope) in the ASGI trace middleware. What do you think?

Reply:
> hmm is this because fastapi/starlette call back into the app for the subapps?

yes, you basically mount a full fastapi app under another one. so you basically call it as a second asgi request with modified scope (e.g. prefix removed, etc)

one other thing Zach and I just noticed / that needs to be checked. are we double counting metrics in this scenario because we have two `fastapi.request` spans?

Reply:
> one other thing Zach and I just noticed / that needs to be checked. are we double counting metrics in this scenario because we have two fastapi.request spans?

good catch, AFAICT yes we would be.

> if yes, I wonder if we need to mark the span as not measured when it is a sub-app call?

ugh yeah I think we have to check this and not mark the sub-app ones as measured. We can't rely on the service-entry logic since there could be manual instrumented spans before the `fastapi.request` span.

Reply:
I tested it a couple of times and don't find the metrics ever doubling. Send 10 requests, see 10 traces, see trace.fastapi.request.hits at 10. I remember now testing this before as well for a Node issue and it not causing a problem. I'll get confirmation of the expected behavior though.

Reply:
It doesn't seem like we mark these as measured. so nothing needs to change right now

if we ever do want/need to mark the `*.request` "root" spans as measured, we will want to be sure to not mark the sub-app ones.